// @lexical/utils/LexicalUtils.esm.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as modDev from './LexicalUtils.dev.esm.js';
import * as modProd from './LexicalUtils.prod.esm.js';
const mod = process.env.NODE_ENV === 'development' ? modDev : modProd;
export const $dfs = mod.$dfs;
export const $filter = mod.$filter;
export const $findMatchingParent = mod.$findMatchingParent;
export const $getNearestBlockElementAncestorOrThrow = mod.$getNearestBlockElementAncestorOrThrow;
export const $getNearestNodeOfType = mod.$getNearestNodeOfType;
export const $insertFirst = mod.$insertFirst;
export const $insertNodeToNearestRoot = mod.$insertNodeToNearestRoot;
export const $restoreEditorState = mod.$restoreEditorState;
export const $splitNode = mod.$splitNode;
export const $wrapNodeInElement = mod.$wrapNodeInElement;
export const addClassNamesToElement = mod.addClassNamesToElement;
export const isHTMLAnchorElement = mod.isHTMLAnchorElement;
export const isHTMLElement = mod.isHTMLElement;
export const isMimeType = mod.isMimeType;
export const markSelection = mod.markSelection;
export const mediaFileReader = mod.mediaFileReader;
export const mergeRegister = mod.mergeRegister;
export const objectKlassEquals = mod.objectKlassEquals;
export const positionNodeOnRange = mod.positionNodeOnRange;
export const registerNestedElementResolver = mod.registerNestedElementResolver;
export const removeClassNamesFromElement = mod.removeClassNamesFromElement;
